### PR TITLE
Remove and update extensions to reflect new stub file

### PR DESCRIPTION
**Pull request by IanDelMar**
Removes the following extensions/rules, which are now fully covered by the new (yet unreleased) stub file:
* GetTaxonomiesDynamicFunctionReturnTypeExtension
* GetObjectTaxonomiesDynamicFunctionReturnTypeExtension
* HasFilterDynamicFunctionReturnTypeExtension
* MySQL2DateDynamicFunctionReturnTypeExtension
* CurrentTimeDynamicFunctionReturnTypeExtension
* WpThemeGetDynamicMethodReturnTypeExtension
* GetPermalinkDynamicFunctionReturnTypeExtension
* WPErrorParameterDynamicFunctionReturnTypeExtension
* IsWpErrorFunctionTypeSpecifyingExtension
* IsWpErrorRule
* EchoParameterDynamicFunctionReturnTypeExtension

Also removes the tests as they should be part of php-stubs/wordpress-stubs itself.

Updates and merges the following extension, which are partly (though almost fully) covered by the new (yet unreleased) stub file:
* GetCommentDynamicFunctionReturnTypeExtension
* GetPostDynamicFunctionReturnTypeExtension

Also removes the tests not related to the return type extension as they should be part of php-stubs/wordpress-stubs itself.

Todo:
- [ ] Update the php-stubs/wordpress-stubs version in composer.json.

Part of #181

For testing purposes I recommend to 
```diff
    "require": {
        "php": "^7.2 || ^8.0",
-        "php-stubs/wordpress-stubs": "^4.7 || ^5.0 || ^6.0",
+        "php-stubs/wordpress-stubs": "dev-master",
        "phpstan/phpstan": "^1.10.0",
        "symfony/polyfill-php73": "^1.12.0"
    },
```
in composer.json

### CODE REVIEW

**Comment by herndlm** [comment]
Not sure about the test removal tbh, is it possible to add them to the stubs package or would they just be removed for now? The latter feels like a risky loss

**Comment by johnbillion** [comment]
Yeah loss of the tests is very unfortunate. Without the return type extension classes, I'm not sure how you'd test the behaviour.

**Comment by szepeviktor** [comment]
What wonder do we have here?? :)

**Comment by herndlm** [comment]
What about dev-requiring phpstan + extension in the stubs repo and making composer aware of the fact that it should use the local stubs checkout instead of the latest release. And the we essentially "just" move the tests over. In theory at least. I can give it a try

**Comment by szepeviktor** [comment]
I have no clue how to manage this dependency problem. For me it takes a couple of months to research.
🤔 Monorepo?

Let's leave tests here until a robust theory comes up.

**Comment by IanDelMar** [comment]
I find it weird to test something that another repository does.

**Comment by szepeviktor** [comment]
> I find it weird to test something that another repository does.

What is the right response here?? "I find it weird to use WordPress" ?

**Comment by szepeviktor** [comment]
The 1st obstacle: How to send a PR that fixes something in the PHPStan extension **and** in core stubs?

**Comment by herndlm** [comment]
I did something in https://github.com/php-stubs/wordpress-stubs/pull/76, it didn't work as I initially planned, but it should be even a bit simpler

**Comment by IanDelMar** [comment]
All tests have been moved to [php-stubs/wordpress-stubs](https://github.com/php-stubs/wordpress-stubs/).

**Comment by IanDelMar** [comment]
@szepeviktor I removed all but one function from the `$wp_error` parameter extension. `_set_cron_array()` is left. This function is marked private by WP. It should not be used and there should not need a dynamic return type extension. What do you think?

**Comment by szepeviktor** [comment]
Please remove that private function too.

**Comment by szepeviktor** [comment]
The next logical step would be to exclude all these `__*` functions from WP stubs, but I am sure removing them will set up a bomb.

**Comment by IanDelMar** [comment]
> The next logical step would be to exclude all these `__*` functions from WP stubs,

Do you mean `_*` or `__*`?

**Comment by szepeviktor** [comment]
Touching anything is WP core a **no-go**.
Or else we have to work 1000+ hours.

**Comment by szepeviktor** [comment]
🎈 Fun fact: there are 462 `\sfunction _[^_]` functions/methods.

**Comment by IanDelMar** [comment]
What are `AssertNotWpErrorTypeSpecifyingExtension` and `AssertWpErrorTypeSpecifyingExtension` for?

File: README.md [hunk 1]
```diff
@@ -73,9 +73,8 @@ Please see [WooCommerce Stubs](https://github.com/php-stubs/woocommerce-stubs)
 
 - Makes it possible to run PHPStan on WordPress plugins and themes
 - Loads [`php-stubs/wordpress-stubs`](https://github.com/php-stubs/wordpress-stubs) package
-- Provides dynamic return type extensions for many core functions
+- Provides dynamic return type extensions for functions that are not covered in [`php-stubs/wordpress-stubs`](https://github.com/php-stubs/wordpress-stubs)
```
Comment (szepeviktor) [inline README.md:76]:
```suggestion
- Provides dynamic return type extensions for functions
    that are not covered in [`php-stubs/wordpress-stubs`](https://github.com/php-stubs/wordpress-stubs/blob/master/functionMap.php)
```

**Comment by szepeviktor** [comment]
❓ Will WordPress 7.0 need no viktorstan to be statically analyzed?

**Comment by IanDelMar** [comment]
Status update

### Extension to check
* WpThemeMagicPropertiesClassReflectionExtension
* ShortcodeAttsDynamicFunctionReturnTypeExtension
* AssertWpErrorTypeSpecifyingExtension
* AssertNotWpErrorTypeSpecifyingExtension

### Extension to keep
* ApplyFiltersDynamicFunctionReturnTypeExtension
* EchoKeyDynamicFunctionReturnTypeExtension
* StringOrArrayDynamicFunctionReturnTypeExtension
* GetTermsDynamicFunctionReturnTypeExtension
* GetPostDynamicFunctionReturnTypeExtension
* GetPostsDynamicFunctionReturnTypeExtension
* GetSitesDynamicFunctionReturnTypeExtension
* WpParseUrlFunctionDynamicReturnTypeExtension

### Extension to discuss
* GetListTableDynamicFunctionReturnTypeExtension 
* RedirectCanonicalDynamicFunctionReturnTypeExtension

#### GetListTableDynamicFunctionReturnTypeExtension
Here's what WP says about `_get_list_table()`: "This is a private function, however, and should not be used by developers." see [here](https://developer.wordpress.org/reference/classes/wp_list_table/). I suggest to remove the extension.

#### RedirectCanonicalDynamicFunctionReturnTypeExtension
This extension fixes `redirect_canonical()` which according to WP returns `string|void` by letting it return `string|null` instead ignoring the `return;`. This should be fixed in core not in an extension. The same goes for functions handled via the echo parameter (already moved to wordpress-stubs) and echo key extension and surely for many more core function that are not covered by phpstan-wordpress and wordpress-stubs. By returning null instead of void `RedirectCanonicalDynamicFunctionReturnTypeExtension` treats this differently than the echo extension. I think how this is treated should be consistent across extension and I suggest to remove `RedirectCanonicalDynamicFunctionReturnTypeExtension`.

**Comment by IanDelMar** [comment]
Thank you @swissspidy

**Comment by szepeviktor** [comment]
@IanDelMar Should I cherry-pick #183 #184 and #185 into 2.x branch?

**Comment by szepeviktor** [comment]
@IanDelMar I think this PR is complete.

Do we still need `"php-stubs/wordpress-stubs": "dev-master"`?

**Comment by IanDelMar** [comment]
> I think this PR is complete.

I agree with it.

> Do we still need "php-stubs/wordpress-stubs": "dev-master"?

Don't think so.

**Pull request merged**